### PR TITLE
rate limit auto-tuning

### PR DESCRIPTION
**Pull request by ajkr**
try a simple demand-based dynamic rate limiter. as commented, it adjusts rate limit upwards when demand is consistently high, and downwards when demand is consistently low. what we really should do is estimate the rate needed to prevent L0 from filling up, but let's try something even simpler for the first stab.

Test Plan: tried in db_bench and watched it increase/decrease rate limit. need help coming up with an evaluation plan

### CODE REVIEW

**Comment by siying** [comment]
Is it ready for code review?

**Comment by ajkr** [comment]
Thanks for checking. I think let's deprioritize this one as I don't have a use case yet or intent to benchmark it in the near future. We can focus on https://github.com/facebook/rocksdb/pull/2027 instead as it seems more likely to help write stall issues.

**Comment by ajkr** [comment]
@siying, do you mind looking at the API for configuring auto-tuning objects and giving any feedback? I plan to play with the rate limit calculation a bit, but the other stuff in this diff might be good to review earlier.

File: utilities/auto_tuners/rate_limiter_auto_tuner.cc [hunk 1]
```diff
@@ -0,0 +1,109 @@
+//  Copyright (c) 2017-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under the BSD-style license found in the
+//  LICENSE file in the root directory of this source tree. An additional grant
+//  of patent rights can be found in the PATENTS file in the same directory.
```
Comment (sagar0) [inline utilities/auto_tuners/rate_limiter_auto_tuner.cc]: You might also need to add a few more lines here like the ones added in #2226 .

Comment (ajkr) [inline utilities/auto_tuners/rate_limiter_auto_tuner.cc]: thanks, done :)

**Comment by ajkr** [comment]
replaced with #2899

**Pull request closed without merging**